src/Stratis.Bitcoin.Features.Consensus/CoinViews/CachedCoinView.cs

```diff
@@ -237,7 +238,7 @@ public async Task<FetchCoinsResponse> FetchCoinsAsync(uint256[] txIds, Cancellat
                     var cache = new CacheItem();
                     cache.ExistInInner = unspent != null;
                     cache.IsDirty = false;
-                    cache.UnspentOutputs = unspent;
+                    cache.UnspentOutputs = unspent?.Clone();
```

Comment:
Ah ok I don't think this is the correct fix, it requires fetching the coins later from disk as @fassadlr mentions.
Cloning at the UTXO set level is more correct in my opinion.

Reply:
Ah ok this does not require fetching again however not sure it addresses the issue of the cache item directly being manipulated when we modify the coinview set in the rules engine.

Reply:
It does, because the result of the call which goes to the coin view rule set is now different instance. So you have one thing going to the cache and another one being returned from the call which is used by the rule

Reply:
Currently the only modification of the cache item should happen in the save change async, so there should not be anything else touching the cache item